jms/src/test/scala/akka/stream/alpakka/jms/scaladsl/JmsAckConnectorsSpec.scala

```diff
@@ -0,0 +1,371 @@
+      val jmsTopicSink: Sink[String, NotUsed] = JmsSink.textSink(
+        JmsSinkSettings(connectionFactory).withTopic("topic")
+      )
+      //#create-topic-sink
+      val jmsTopicSink2: Sink[String, NotUsed] = JmsSink.textSink(
+        JmsSinkSettings(connectionFactory).withTopic("topic")
+      )
+
+      val in = List("a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k")
+      val inNumbers = (1 to 10).map(_.toString)
+
+      //#create-topic-source
+      val jmsTopicSource: Source[AckEnvelope, KillSwitch] = JmsSource.ackSource(
+        JmsSourceSettings(connectionFactory).withSessionCount(1).withBufferSize(0).withTopic("topic")
+      )
+      //#create-topic-source
+      val jmsSource2: Source[AckEnvelope, KillSwitch] = JmsSource.ackSource(
+        JmsSourceSettings(connectionFactory).withSessionCount(1).withBufferSize(0).withTopic("topic")
+      )
+
+      val expectedSize = in.size + inNumbers.size
+      //#run-topic-source
+      val result1 = jmsTopicSource
+        .take(expectedSize)
+        .map(env => (env, env.message.asInstanceOf[TextMessage].getText))
+        .map { case (env, text) => env.acknowledge(); text }
+        .runWith(Sink.seq)
+        .map(_.sorted)
+      val result2 = jmsSource2
+        .take(expectedSize)
+        .map(env => (env, env.message.asInstanceOf[TextMessage].getText))
+        .map { case (env, text) => env.acknowledge(); text }
+        .runWith(Sink.seq)
+        .map(_.sorted)
+      //#run-topic-source
+
+      //We wait a little to be sure that the source is connected
+      Thread.sleep(500)
+
+      //#run-topic-sink
+      Source(in).runWith(jmsTopicSink)
+      //#run-topic-sink
+      Source(inNumbers).runWith(jmsTopicSink2)
+
+      val expectedList: List[String] = in ++ inNumbers
+      result1.futureValue should contain theSameElementsAs expectedList
+      result2.futureValue should contain theSameElementsAs expectedList
+    }
+
+    "ensure no message loss when stopping a stream" in withServer() { ctx =>
+      val url: String = ctx.url
+      val connectionFactory = new ActiveMQConnectionFactory(url)
+
+      val jmsSink: Sink[JmsTextMessage, NotUsed] = JmsSink(
+        JmsSinkSettings(connectionFactory).withQueue("numbers")
+      )
+
+      val numsIn = 1 to 100
+      val msgsIn = numsIn map { n =>
+        JmsTextMessage(n.toString).add("Number", n)
+      }
+
+      Source(msgsIn).runWith(jmsSink)
+
+      val jmsSource: Source[AckEnvelope, KillSwitch] = JmsSource.ackSource(
+        JmsSourceSettings(connectionFactory).withSessionCount(5).withBufferSize(0).withQueue("numbers")
+      )
+
+      val resultQueue = new LinkedBlockingQueue[String]()
+
+      val killSwitch = jmsSource
+        .take(msgsIn.size)
+        .throttle(10, 1.second, 2, ThrottleMode.shaping)
+        .map { env =>
+          resultQueue.add(env.message.asInstanceOf[TextMessage].getText)
+          env.acknowledge()
+        }
+        .to(Sink.ignore)
+        .run()
```

Comment:
Entire part from `map` to `run` can be a single `runForeach` as all it does is side effect (same fora few more identical below)

Reply:
I'm fixing this where I can. Since `runForeach` does not give me the `Source`s materialized value, I'll have to stick to this for the tests I need the `KillSwitch` back out. But if there are better suggestions, please do let me know.

Reply:
Ah, didn't think of that, then I think `.to(Sink.foreach { side-effecting })` would make it more clear than mapping elements to `Unit`